### PR TITLE
Fix yet another epoch slot issue

### LINKED ISSUE #101: AssertionError in get_crosslink_committees_at_slot after slot 64
On Linux x86-64

Similar to #89 #97 we have an issue during epoch transition, this times after 64 slots, for epoch 144115188075855873 

Full logs available in https://gist.github.com/mratsim/a6539c049dfb49a638327b9db5cec44a.
There are 2 logs, one with multiple nodes desynced, one with only a single node desynced.

The following log is after merging #103 for epochs printing.

```
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [AssertionError]
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=338 slot=121 validator=96d7e599 node=0
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=296 slot=114 validator=8e724e86 node=5
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [AssertionError]
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=326 slot=119 validator=0a89ffbe node=2
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [AssertionError]
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [AssertionError]
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=350 slot=123 validator=91792a76 node=7
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [AssertionError]
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [Assert
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [Assert
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=344 slot=122 validator=842ef0e2 node=8
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=356 slot=124 validator=861b6b0b node=8
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=362 slot=125 validator=928a8a8e node=8
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=374 slot=127 validator=9090c4be node=1
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [Assert
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [Assert
INF 2019-02-11 16:09:08+01:00 Scheduling block proposal                  thread=0 fromNow=380 slot=128 validator=968a8996 node=8
beacon_node.nim(435)     beacon_node
beacon_node.nim(399)     processBlocks
asyncloop.nim(756)       runForever
asyncloop.nim(238)       poll
beacon_node.nim(328)     :anonymous
beacon_node.nim(306)     scheduleEpochActions
validator.nim(95)        get_crosslink_committees_at_slot
system.nim(3879)         failedAssertImpl
system.nim(3872)         raiseAssert
system.nim(2918)         sysFatal
Error: unhandled exception: /home/beta/Programming/Status/nim-beacon-chain/beacon_chain/spec/validator.nim(95, 10) `epoch <= next_epoch` Previous epoch: 144115188075855872, epoch: 144115188075855874, Next epoch: 144115188075855873 [Assert
Terminated
```

### PR DESCRIPTION
**Pull request by mratsim**
This fixes #101 and add several more tracing, the gist of it is that we were making our updates on a copy of our state.

It does not seem like the copy is needed anymore (pending CI).

In summary I think that this goes one step beyond the branch https://github.com/status-im/nim-beacon-chain/tree/yglukhov-somewhat-stable-simulation

I definitely felt the tangle while trying to debug it (cf: https://github.com/status-im/nim-beacon-chain/pull/117#issuecomment-464746588). I think we need a refactoring as soon as phase 0 stabilizes.

## Next steps

Note that we still have block slot issues, but the new one seems to need moving or reordering core part of the processing as we:
  -  proposed a block
  - received our own proposal a couple of slots later
  - crash with unexpected block slot.

deepinscreenshot_select-area_20190219163745

With the assert being at the soon infamous :) "Could this fail somehow" 

https://github.com/status-im/nim-beacon-chain/blob/e1efdd4349a35903861a5384df0a96f07005fa56/beacon_chain/beacon_node.nim#L220-L229

### CODE REVIEW

File: beacon_chain/state_transition.nim [hunk 1]
```diff
@@ -370,9 +370,11 @@ proc processBlock(
 
   # https://github.com/ethereum/eth2.0-specs/blob/v0.2.0/specs/core/0_beacon-chain.md#slot-1
   if not (blck.slot == state.slot):
-    notice "Unexpected block slot number",
-      blockSlot = blck.slot,
-      stateSlot = state.slot
+    notice "Unexpected block slot number - stackTrace incoming",
+      blockSlot = humaneSlotNum blck.slot,
+      stateSlot = humaneSlotNum state.slot
+    writeStackTrace()
```
Comment (arnetheduck) [inline beacon_chain/state_transition.nim]: remove debugging cruft?

File: beacon_chain/state_transition.nim [hunk 2]
```diff
@@ -397,23 +399,29 @@ proc processBlock(
       return false
 
   if not processRandao(state, blck, flags):
+    notice "Randao failure"
```
Comment (arnetheduck) [inline beacon_chain/state_transition.nim]: these should already log at the error site, with more pertinent information (where the initial error is detected)

File: beacon_chain/state_transition.nim [hunk 3]
```diff
@@ -804,6 +812,12 @@ proc updateState*(state: var BeaconState, previous_block_root: Eth2Digest,
   #      bool return values...)
   # TODO There's a discussion about what this function should do, and when:
   #      https://github.com/ethereum/eth2.0-specs/issues/284
+  
+  debug "TRACE - updateState",
+    oldStateSlot = humaneSlotNum state.slot,
+    proposedBlockSlot = if new_block.isSome(): new_block.get.slot.humaneSlotNum
+                        else: 1010101010
```
Comment (arnetheduck) [inline beacon_chain/state_transition.nim]: what?

File: beacon_chain/beacon_node.nim [hunk 4]
```diff
@@ -329,11 +340,19 @@ proc scheduleEpochActions(node: BeaconNode, epoch: uint64) =
...
+    $humaneEpochNum(node.beaconState.get_current_epoch)
+
   info "Scheduling next epoch update",
     fromNow = (at - fastEpochTime()) div 1000,
     epoch = humaneEpochNum(nextEpoch)
 
+  debug "TRACE - before closure: currentSlot ", stateSlot = humaneSlotNum node.beaconState.slot
   addTimer(at) do (p: pointer):
+    # Chronicles is not GC-safe
+    debugEcho "TRACE - in closure: currentSlot ", humaneSlotNum node.beaconState.slot
```
Comment (arnetheduck) [inline beacon_chain/beacon_node.nim]: ..

**Comment by arnetheduck** [comment]
#125 probably fixes this issue

**Comment by mratsim** [comment]
Actually I wanted to get more input from @zah to make sure the copy was still needed, hence I left the debug output so that he can reproduce and check.

**Pull request closed without merging**